server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,436 @@
+ * <ul>
+ * <li>Remove cluster-manager eligible nodes from voting config [TODO - checks to avoid quorum loss scenarios]</li>
+ * <li>Initiates nodes decommissioning by adding custom metadata with the attribute and state as {@link DecommissionStatus#DECOMMISSION_INIT}</li>
+ * <li>Triggers weigh away for nodes having given awareness attribute to drain. This marks the decommission status as {@link DecommissionStatus#DECOMMISSION_IN_PROGRESS}</li>
+ * <li>Once weighed away, the service triggers nodes decommission</li>
+ * <li>Once the decommission is successful, the service clears the voting config and marks the status as {@link DecommissionStatus#DECOMMISSION_SUCCESSFUL}</li>
+ * <li>If service fails at any step, it would mark the status as {@link DecommissionStatus#DECOMMISSION_FAILED}</li>
+ * </ul>
+ *
+ * @opensearch.internal
+ */
+public class DecommissionService {
+
+    private static final Logger logger = LogManager.getLogger(DecommissionService.class);
+
+    private final ClusterService clusterService;
+    private final TransportService transportService;
+    private final ThreadPool threadPool;
+    private final DecommissionHelper decommissionHelper;
+    private volatile List<String> awarenessAttributes;
+
+    @Inject
+    public DecommissionService(
+        Settings settings,
+        ClusterSettings clusterSettings,
+        ClusterService clusterService,
+        TransportService transportService,
+        ThreadPool threadPool,
+        AllocationService allocationService
+    ) {
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.threadPool = threadPool;
+        this.decommissionHelper = new DecommissionHelper(
+            clusterService,
+            allocationService,
+            threadPool
+        );
+        this.awarenessAttributes = AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(
+            AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING,
+            this::setAwarenessAttributes
+        );
+    }
+
+    List<String> getAwarenessAttributes() {
+        return awarenessAttributes;
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    public void initiateAttributeDecommissioning(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener,
+        ClusterState state
+    ) {
+        validateAwarenessAttribute(decommissionAttribute, getAwarenessAttributes());
+        logger.info("initiating awareness attribute [{}] decommissioning", decommissionAttribute.toString());
+        excludeDecommissionedClusterManagerNodesFromVotingConfig(decommissionAttribute);
+        registerDecommissionAttribute(decommissionAttribute, listener);
```

Comment:
There are 2 different steps and still wont guarantee atomicity and consistency. Lets consider this -> ```excludeDecommissionedClusterManagerNodesFromVotingConfig``` executed and before ```registerDecommissionAttribute``` this could be executed, user removed the voting exclusion (as that is an external API). This would leave the system in inconsistent state where decommissioned node could become master.

Reply:
Updated the code in latest commit. Now, the we will register the attribute (also the initiation of decommissioning only when the local is not going to be decommissioned and is active master. All we need here, the request doesn't run on to be decommissioned master node. Snippet - 
```
// remove all decommissioned cluster manager eligible nodes from voting config
        // The method ensures that we don't exclude same nodes multiple times
        excludeDecommissionedClusterManagerNodesFromVotingConfig(decommissionAttribute);

        // explicitly throwing NotClusterManagerException as we can certainly say the local cluster manager node will
        // be abdicated and soon will no longer be cluster manager.
        if(transportService.getLocalNode().isClusterManagerNode()
            && !nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) {
            registerDecommissionAttribute(decommissionAttribute, listener);
        } else {
            throw new NotClusterManagerException(
                "node ["
                    + transportService.getLocalNode().toString()
                    + "] not eligible to execute decommission request. Will retry until timeout."
            );
        }
```

Reply:
Do you suggest any other change so a decommissioned node is never elected as a master again? Should we add checks in `LeaderChecker`?

Reply:
you need this check 
```if(transportService.getLocalNode().isClusterManagerNode()
            && !nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) { 
```
 in the cluster state submit task in registerDecommisionAttribute method to ensure active master doesnt have decommission attribute. this check here wont prevent the race condition completely.

Reply:
Adding this check in submit state update call would rather cause the transport call to end with failure but here I need to retry the call if the current is master but will be decommissioned. Also, even if there's a race condition, the else part will ensure the transport call is retried. Adding the updated code below. Please do let me know if there's a miss in my understanding.

```
// explicitly calling listener.onFailure with NotClusterManagerException as we can certainly say the local cluster manager node will
        // be abdicated and soon will no longer be cluster manager.
        if (transportService.getLocalNode().isClusterManagerNode()
            && !nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)) {
            // we are good here to send the response now as the request is processed by an eligible active leader
            listener.onResponse(new ClusterStateUpdateResponse(true));
            failDecommissionedNodes(clusterService.getClusterApplierService().state());
        } else {
            // this will ensure that request is retried until cluster manager times out
            listener.onFailure(new NotClusterManagerException("node ["
                + transportService.getLocalNode().toString()
                + "] not eligible to execute decommission request. Will retry until timeout."));
        }
```